### PR TITLE
firmware & tools: chamoc tools and chamos module implementation 

### PR DESCRIPTION
**Pull request by eduazocar**
### Contribution description

This is an implementation of vaina source code, The objective of this PR is design an udp client to connect with an device and send it a NIB_ADD request or a NIB_DEL, this device will response with an ACK if the request was received correctly, if the request was wrong received, the device should response with an  NACK.

These Pr contains some structures to gives the elements required to this process, refering an client by an interface.

The interface that could be used are that appears in your System. you could see these executing `ip address show` command or `ifconfig`
```sh
enp2s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.2.119  netmask 255.255.255.0  broadcast 192.168.2.255
        inet6 fe80::bef4:bb4a:f786:e805  prefixlen 64  scopeid 0x20<link>
...
enx5a008df2d56b: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.199  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fe80::98eb:d858:51a0:dcc0  prefixlen 64  scopeid 0x20<link>
        ether 5a:00:8d:f2:d5:6b  txqueuelen 1000  (Ethernet)
...
lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
...
```
 
 

### Testing procedure

There are two cases to test the client connection, first testing locally by the an local host and then connecting with an m4a-24g board.

### 1 - Case °1 Running with chamoc_server test and the chamoc tool.

follow these steps:
- open two terminals (Terminal A & Terminal B):
```sh
# Terminal A
cd test/chamoc_server/
./run.sh
```
The script will ask about which interface will be used 
when you see this output 
```sh
# Terminal A
Server it's running
```
means that you are running the chamoc server, this will nerver stop until that received and nib request from a chamoc client.

- start the client code to local host test:
*note:* before to start run the client, it's recommended take note of the iface name (see typing `ip address show` or `ifconfig`).
in terminal B go to dist/tools/chamoc folder and ther run the test script.
```sh
# Terminal B
cd dist/tools/chamoc
./test.sh
```
When you are started you need to fill all the params fields that the script will need to execute te client code, e.g:
```sh
# Terminal B
Interface: enp2s0
Ipv6 Address: 2001:db8::1
Ipv6 Address prefix (64 - 128): 64
```

your expected output should see just like this:
```sh
# Terminal A
Server it's running
Packet Received:02004020010db8000000000000000000000001
```

```sh
# Terminal B
running nib add request
Request Confirmed
```
In terminal A it can look the packet received in hexadecimals pais,  where:
#### Packet representation
| 02| 00| 40| 20010db8000000000000000000000001|
|:----------------:|:---------------:|:--------------:|:-----------------------------------------------------:|
| type of message | Message sequence number |  network prefix of the ipv6 address | Ipv6 address |

| Packet HEX   |   Description |
|:------------------:| :------------------: |
|  02 | Refers to the NIB type of message (2) is NIB_ADD,  (3) NIB_DEL |
|  00 | Refers to Message sequence number |
|  40 | It's the network prefix of the ipv6 address (`0x40` is 64 in decimal value) |
|  20010db8000000000000000000000001| it's the entered address 2001:db8:1 |

### 2 - Case °2 Running with chamos_server in m4a-24g board

follow these steps:
- open two terminals (Terminal A & Terminal B):
```sh
# terminal A
cd dist/tools/chamoc
./autoconfigure.sh
```
*note* : you could run the autoconfig script placing your owns parameters to the nib message and setup of the host, 
`./autoconfigure.sh` `[ipv6 address]`
When you execute the script without any input, by default the script will take the ipv6 address 2001:db8::1 by default.

### CODE REVIEW

File: dist/tools/socket_client/socket_msg.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * @author      eduazocar <eduardoazocar7@gmail.com>
+ */
+
+#ifndef SOCKET_MSG_H
+#define SOCKET_MSG_H
+
+#define MSG_NIB_ADD 1
+#define MSG_NIB_DEL 2
+#define MSG_ACK 3
+#define MSG_NACK 4
```
Comment (luisan00) [inline dist/tools/socket_client/socket_msg.h:40]: For a better understanding this should be an `enum`

**Review by luisan00 — CHANGES_REQUESTED** [review]
1.- The tool name is ambiguous 
2.- What's the purpose of a command in this tool?

File: firmware/network/chamos/include/chamos.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#endif
+
+/**
+ * @enum   CHAMOS message types
+ */
+enum chamos_msg_types_t{
+    MSG_ACK = 0,  /**< Message acknowledged */
+    MSG_NACK = 1, /**< Message not acknowledged */
+    MSG_NIB_ADD = 2,  /**< Add entry to NIB */
+    MSG_NIB_DEL = 3,  /**< Delete entry from NIB */
```
Comment (luisan00) [inline firmware/network/chamos/include/chamos.h:43]: wrong inline documentation format

File: firmware/network/chamos/include/chamos.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+/**
+ * @brief   CHAMOS message
+ */
+typedef struct {
+    uint8_t msg_type; /**< Message type */
+    uint8_t seqno; /**< Sequence number */
+    uint8_t ip_len; /**< IP address pfx_len */
+    ipv6_addr_t ip;  /**< IP address */
```
Comment (luisan00) [inline firmware/network/chamos/include/chamos.h:53]: wrong inline documentation format

**Review by luisan00 — CHANGES_REQUESTED** [review]
nice job !! :100: 
just documentation and aesthetics by now

File: firmware/network/chamos/include/chamos.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ * @param good_ack  [in]      bool
+ * @return int
+ */
+int server_send_ack(chamos_msg_t *msg, sock_udp_ep_t *remote, bool flag_ack_val);
+
+/**
+ * @brief this function is used to init the module
+ *
+ * @param netiface       nteif
+ * @param port     [in]  port
```
Comment (eduazocar) [inline firmware/network/chamos/include/chamos.h:70]: Fix format nteif to netif

**Review by eduazocar — COMMENTED** [review]
Good, only has a little doc corrections and format corrections :+1:

File: dist/tools/chamoc/chamoc_msg.c [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                printf("Bad Request, Trying Again\n");
+            } else {
+                printf("Time Out\n\n");
+            }
+        }
+    }
+    return 1;
+}
+
+uint8_t craft_seqno(chamos_client_t *client) { client->seqno += 1; }
```
[the remark this answers is not in the document]
Comment (eduazocar) [reply dist/tools/chamoc/chamoc_msg.c:69]: Sorry that was generated by the clang-format, i will fixed it

File: dist/tools/chamoc/doc.txt [hunk 6]
```diff
@@ -0,0 +1,102 @@
+/**
+@defgroup chamoc Chamoc
+@ingroup  network
```
Comment (xkevin190) [inline dist/tools/chamoc/doc.txt:3]: this module is not part of the firmware, please add one correct group.

**Review by xkevin190 — CHANGES_REQUESTED** [review]

File: dist/tools/chamoc/chamoc_msg.c [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include "client.h"
+#include "chamoc_msg.h"
+
+int serialize(chamoc_message_t msg, char *buffer, uint8_t len) {
+    uint8_t strbuff[19] = {msg.message, msg.seqno, msg.prefix};
```
Comment (xkevin190) [inline dist/tools/chamoc/chamoc_msg.c:34]:
```suggestion
    uint8_t strbuff[sizeof(chamoc_message_t)] = {msg.message, msg.seqno, msg.prefix};
```

File: dist/tools/chamoc/nib.c [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+#include <stdio.h>
+#include <string.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#include "client.h"
+#include "chamoc_msg.h"
+
+int nib_req(char *iface, char *ip, uint8_t prefix, uint8_t msg_type) {
+    if (msg_type == 2 || 3) {
```
[the remark this answers is not in the document]
Comment (eduazocar) [reply dist/tools/chamoc/nib.c:32]: It's the same

File: firmware/network/chamos/chamos.c [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    switch (type) {
+    case MSG_NIB_ADD:
+        if (len < sizeof(chamos_msg_t)) {
+            printf("Chamos: NIB ADD invalid message size! \n");
+            return -EINVAL;
+        }
+        chamos_msg->msg_type = type;
+        chamos_msg->seqno = seqno;
+        chamos_msg->ip_len = ip_len;
+        memcpy(&chamos_msg->ip, &buffer[3], sizeof(ipv6_addr_t));
```
Comment (xkevin190) [inline firmware/network/chamos/chamos.c:95]: verify the address is correct before doing this

File: firmware/network/chamos/chamos.c [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        break;
+    case MSG_NIB_DEL:
+        if (len < sizeof(chamos_msg_t)) {
+            printf("Chamos: NIB DEL invalid message size! \n");
+            return -EINVAL;
+        }
+        chamos_msg->msg_type = type;
+        chamos_msg->seqno = seqno;
+        chamos_msg->ip_len = ip_len;
+        memcpy(&chamos_msg->ip, &buffer[3], sizeof(ipv6_addr_t));
```
Comment (xkevin190) [inline firmware/network/chamos/chamos.c:105]: verify the address is correct before doing this

**Review by xkevin190 — CHANGES_REQUESTED** [review]

File: dist/tools/chamoc/start_network.sh [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    ${SUDO} ip route del "${PREFIX}" via  2001:db8::1 dev "${INTERFACE}"
+}
+
+cleanup() {
+    echo "Cleaning up..."
+    cleanup_interface
+    trap "" INT QUIT TERM EXIT
+}
+
+
```
Comment (luisan00) [inline dist/tools/chamoc/start_network.sh:52]:
```suggestion
```

File: tests/chamoc_server/run.sh [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    ./${TEST_APP} ${INTERFACE} ${SERVER_PORT}
+}
+
+read -p "Interface: " -r INTERFACE
+if [ -z "${INTERFACE}" ]; then
+    echo "Any parameter cannot be empty"
+else
+compile
+run
+fi
```
Comment (luisan00) [inline tests/chamoc_server/run.sh:26]:
```suggestion
fi

```

**Review by luisan00 — CHANGES_REQUESTED** [review]
well done. almost finished!

**Review by xkevin190 — APPROVED** [review]

**Review by luisan00 — APPROVED** [review]
well dona!! going ahead!! :+1:

**Pull request merged**